util/dvsim/Scheduler.py

```diff
@@ -48,15 +46,23 @@ def __init__(self, items):
         # _queued is a list so that we dispatch things in order (relevant
         # for things like tests where we have ordered things cleverly to
         # try to see failures early). They are maintained for each target.
+
+        # The list of available targets and the list of running items in each
+        # target are maintained as CircularLists, so that they can be iterated
+        # over indefinitely. This allows us to poll a smaller subset of jobs
+        # rather than the entire regression. We keep rotating through our list
+        # of running items, picking up where we left off on the last poll.
         self._queued = {}
         self._running = {}
         self._passed = {}
         self._failed = {}
         self._killed = {}
         self._total = {}
+        self._targets = CircularList()
+        self._targets.extend(self._scheduled.keys())
```

Comment:
This is fine, I guess, but it seems rather complicated. If we only need a circular list in one place, wouldn't it be easier just to store an offset and then do the "grab an interval, wrapping around if necessary" logic in a helper method? I'd have thought that would be ~10 lines of code, rather than the 142 in the helper class.

Reply:
i thought about that as well, but realized that the hard part is grabbing N objects across multiple running targets which would have ended up being a bit more complicated. Plus, we are saving time creating and destroying lists (`to_pass` and `to_fail`) and iterating over those (2 extra loops) every second we poll with this approach (allows our list to be mutated while circularly iterating over it). The current implementation is actually a bit faster especially when running over 1000s of tests.  

IMO a circular linked list is a commonly used data structure. If you look at the implementation, it is fairly easy to follow. Agree that it is a lot of lines, but its mostly to make it resemble a `list` as much as possible. Turns out there is a 3rd party library implementation that I could reuse: https://pypi.org/project/dstructure/
I can update the code to reuse that if you prefer.

Reply:
Hmm, ok. I don't think I'd add an external dependency for this (and `dstructure` looks really young with no obvious repository: probably not a great dependency to gain!)